### PR TITLE
Feature/platfowner/function permission

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Use function id in function permission setting in rule configs (auth.fid vs auth.addr) 
- Add function permission in genesis rules and test rules
- Add tests for function permission (with _saveLastTx() native function) 

#195

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -781,16 +798,13 @@ class P2pServer {
               PredefinedDbPaths.SHARDING_SHARD,
               ainUtil.encode(shardingPath)
             ]),
-            value: GenesisSharding
+            value: genesisSharding
```
Comment (liayoo) [inline server/index.js:801]: I think this is still capitalized?

File: db/functions.js [hunk 2]
```diff
@@ -193,20 +196,86 @@ class Functions {
...
+    const execResult = Functions.buildExecutionResult(timestamp, transaction.hash, code);
+    return this.setValueOrLog(resultPath, execResult, auth, timestamp);
+  }
+
+  /**
+   * Saves the transaction's hash to a sibling path.
+   * e.g.) For tx's value path 'path/to/value', it saves the tx hash to 'path/to/last_tx'
+   */
+  // NOTE(seo): Use this function for test-purposes.
+  _saveLastTx(value, context) {
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/functions.js:247]: Just tried to generalize its functionality to make it more useful. PTAL

File: server/index.js (hunk 1 above)
Comment (platfowner) [reply server/index.js]: Just removed redundant function params.

**Comment by platfowner** [comment]
Comments addressed. PTAL! @liayoo

File: db/functions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply db/functions.js:247]: Looks great!!

**Comment by liayoo** [comment]
LGTM. Thanks for the new feature! We now have more fine-grained permissions 👍

**Review by liayoo — APPROVED** [review]

**Comment by platfowner** [comment]
Thanks for the fast review. Let me merge this now without integration testing (I just checked it passes the test before the p2p PR). @liayoo

**Pull request merged**